// frontend/src/components/Calendar/Calendar.jsx
import Button from '../ui/Button/Button';
import SearchView from '../ui/SearchView/SearchView';
import CalendarSidebar from '../LeftSide/LeftSide';
import BigCalendar from '../BigCalendar/BigCalendar';
import React, { useState } from 'react';
import './Calendar.css';

import WeekView from '../BigCalendar/WeekCalendar'
import DayView from '../BigCalendar/DayCalendar'
import MonthView from '../BigCalendar/MonthCalendar'
import YearView from '../BigCalendar/YearCalendar'

export default function Calendar() {
  const [view, setSelectedView] = useState('Week');
  const [currentDate, setCurrentDate] = useState(new Date());

  const [currentInfo, setCurrentInfo] = useState({
    year: null,
    month: null,
    day: null
  });

  const [calendarId, setCalendarId] = useState(null);
  const [calendars, setCalendars] = useState([]);
  const [allEvents, setAllEvents] = useState([]);
  const [visibleCalendars, setVisibleCalendars] = useState({});
  const [showNewEvent, setShowNewEvent] = useState(false);
  const [selectedEvent, setSelectedEvent] = useState(null);
  const [showEventDetails, setShowEventDetails] = useState(false);
  const [showShareEvent, setShowShareEvent] = useState(false);
  const [showEditEvent, setShowEditEvent] = useState(false);
  const [popupPosition, setPopupPosition] = useState({ x: 200, y: 120 });
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);
  // const [userSettings, setUserSettings] = useState(null);

  // useEffect(() => {
  //   fetch("http://localhost:3000/api/auth/me", { credentials: "include" })
  //     .then(res => res.json())
  //     .then(user => {        
  //       if (user && user._id) {
  //         setUserSettings({
  //           country: user.country,
  //           timeFormat: user.time_format
  //         });
  //       }
  //     })
  //     .catch(() => {});
  // }, []);

  const { settings } = useSettings();

  useEffect(() => {
    const fetchCalendars = async () => {
      try {
        const response = await fetch('http://localhost:3000/api/calendars', {
          credentials: 'include',
          headers: { 'Accept': 'application/json' }
        });
        
        if (!response.ok) {
          throw new Error('Failed to fetch calendars');
        }

        const data = await response.json();
        
        if (data.myCalendars && data.myCalendars.length > 0) {
          setCalendarId(data.myCalendars[0]._id);
        } else {
          // await createDefaultCalendar();
        }
      } catch (error) {
        console.error('Error fetching calendars:', error);
        setError('Failed to load calendars. Please check your connection.');
      }
    };

    fetchCalendars();
  }, []);

 const [user, setUser] = useState(null);

useEffect(() => {
  fetch("http://localhost:3000/api/auth/me", { credentials: "include" })
    .then(res => res.json())
    .then(data => {
      if (data && data._id) {
        setUser(data);
      }
    })
    .catch(() => {});
}, []);


  // фетч ВСЁ
  const fetchEvents = async () => {
    setLoading(true);
    try {
      const response = await fetch('http://localhost:3000/api/calendars', {
        credentials: 'include',
        headers: { 'Accept': 'application/json' }
      });

      if (!response.ok) {
        throw new Error('Failed to fetch calendars');
      }

      const data = await response.json();
      const allCalendars = [...(data.myCalendars || []), ...(data.otherCalendars || [])];
      
      // v tom chisle tsveta
      setCalendars(allCalendars);
      
      const eventsPromises = allCalendars.map(calendar =>
        fetch(`http://localhost:3000/api/events/${calendarId}`, {
          credentials: 'include',
          headers: { 'Accept': 'application/json' }
        })
          .then(res => res.ok ? res.json() : [])
          .then(events => events.map(event => ({ ...event, calendarId: calendar._id })))
          .catch(() => [])
      );

      const eventsArrays = await Promise.all(eventsPromises);
      const allEvents = eventsArrays.flat();
      
      setAllEvents(allEvents);
    } catch (error) {
      console.error('Error fetching events:', error);
      setAllEvents([]);
    } finally {
      setLoading(false);
    }
  };

  useEffect(() => {
    fetchEvents();
  }, []);

  //populate holidays
  useEffect(() => {
    if (!user) return;
    fetch(`http://localhost:3000/api/calendar/${user._id}/${calendarId}/populate-holidays`, {
        method: "POST",
        credentials: "include"
    })
    .then(() => console.log("Holiday events created"))
    .catch(err => console.error(err));
  }, [user]);

  // фильтр видимости
  const visibleEvents = allEvents.filter(event => 
    visibleCalendars[event.calendarId] !== false
  );

  // смена видимости по клику
  const handleCalendarVisibilityChange = (newVisibleCalendars) => {
    setVisibleCalendars(newVisibleCalendars);
  };

  const handleEventCreated = (newEvent) => {
    setAllEvents([...allEvents, newEvent]);
  };

  const handleShareEvent = (event) => {
    setSelectedEvent(event);
    setShowEventDetails(false);
    setShowShareEvent(true);
  };

  const handleEditEvent = (event) => {
    setSelectedEvent(event);
    setShowEventDetails(false);
    setShowEditEvent(true);
  };

  // рефетч данных
  const handleDataCreated = (type, data) => {
    fetchEvents();
  };

  const handleEventDeleted = (eventId) => {
    setAllEvents(allEvents.filter(e => e._id !== eventId));
  };

  const handleEventUpdated = (updatedEvent) => {
    setAllEvents(allEvents.map(e => 
      e._id === updatedEvent._id ? updatedEvent : e
    ));
  };

  const handleSmallCalendarDaySelect = (date) => {
    setCurrentDate(date);
    setSelectedView("Day");

    setCurrentInfo({
      year: date.getFullYear(),
      month: date.getMonth(),
      day: date.getDate(),
    });
  };

  const renderView = () => {
    const commonProps = { onDateChange: setCurrentInfo, currentDate };
    switch (view) {
      case "Day":
        return <DayView {...commonProps} />;
      case "Month":
        return <MonthView {...commonProps} />;
      case "Year":
        return <YearView {...commonProps} />;
      case "Week":
      default:
        return <WeekView {...commonProps} />;
    }
  };

const handlePrev = () => {
  switch(view) {
    case 'Day':
      setCurrentDate(prev => new Date(prev.getFullYear(), prev.getMonth(), prev.getDate() - 1));
      break;
    case 'Week':
      setCurrentDate(prev => new Date(prev.getFullYear(), prev.getMonth(), prev.getDate() - 7));
      break;
    case 'Month':
      setCurrentDate(prev => new Date(prev.getFullYear(), prev.getMonth() - 1, prev.getDate()));
      break;
    case 'Year':
      setCurrentDate(prev => new Date(prev.getFullYear() - 1, prev.getMonth(), prev.getDate()));
      break;
  }
};

const handleNext = () => {
  switch(view) {
    case 'Day':
      setCurrentDate(prev => new Date(prev.getFullYear(), prev.getMonth(), prev.getDate() + 1));
      break;
    case 'Week':
      setCurrentDate(prev => new Date(prev.getFullYear(), prev.getMonth(), prev.getDate() + 7));
      break;
    case 'Month':
      setCurrentDate(prev => new Date(prev.getFullYear(), prev.getMonth() + 1, prev.getDate()));
      break;
    case 'Year':
      setCurrentDate(prev => new Date(prev.getFullYear() + 1, prev.getMonth(), prev.getDate()));
      break;
  }
};


const handleDayClick = (day) => {
  onDateChange(new Date(currentDate.getFullYear(), currentDate.getMonth(), day));
};

  return (
    <>
      {/* <Helmet>
        <title>Weekly Calendar View | Calendar Pro</title>
      </Helmet> */}
        <div className="calendar-main">
          <div className="calendar-layout">
            <CalendarSidebar />

            <div className="calendar-content">

              <div className="calendar-toolbar">
                  <Button text="Today" onClick={() => {}} />
                <div className="toolbar-left">
                  <i className="fa-solid fa-chevron-left calendar-arrow" onClick={() => handlePrev()}></i>
                  <i className="fa-solid fa-chevron-right calendar-arrow" onClick={() => handleNext()}></i>
                </div>
                
                <span>
                  {currentInfo.year && currentInfo.month !== null
                    ? `${currentInfo.day ? currentInfo.day + " " : ""}${currentInfo.month + 1}/${currentInfo.year}`
                    : "Loading..."}
                </span>
                <div className="toolbar-center">
                  <Button text="Day" onClick={() => setSelectedView('Day') } />
                  <Button text="Week" onClick={() => setSelectedView('Week')} />
                  <Button text="Month" onClick={() => setSelectedView('Month')} />
                  <Button text="Year" onClick={() => setSelectedView('Year')} />
                </div>

                <div className="toolbar-right">
                  <SearchView placeholder="Search" />
                </div>
              </div>
              <div className="big-calendar">
                {/* <BigCalendar view={view}> */}
                 <BigCalendar>
                  {renderView()}
                </BigCalendar>
              </div>
            </div>
          </div>
        </div>
    </>
  );
};
